internal/pkg/docker/scheduler/scheduler.go

```diff
@@ -0,0 +1,312 @@
+
+// DockerEngine is used by Scheduler to manage containers.
+type DockerEngine interface {
+	Run(context.Context, *dockerengine.RunOptions) error
+	IsContainerRunning(context.Context, string) (bool, error)
+	Stop(context.Context, string) error
+}
+
+// NewScheduler creates a new Scheduler. idPrefix is a prefix used when
+// naming containers that are run by the Scheduler.
+func NewScheduler(docker DockerEngine, idPrefix string, logOptions logOptionsFunc) *Scheduler {
+	s := &Scheduler{
+		idPrefix:   idPrefix,
+		logOptions: logOptions,
+		stopped:    make(chan struct{}),
+		docker:     docker,
+		wg:         &sync.WaitGroup{},
+		stopOnce:   &sync.Once{},
+		actions:    make(chan action),
+		runErrs:    make(chan error),
+	}
+	return s
+}
+
+// Start starts the Scheduler. Start must be called before any other
+// Scheduler functions. Errors from containers run by the scheduler
+// or from Scheduler actions are sent to the returned error channel.
+// The returned error channel is closed after calling Stop() has
+// stopped the Scheduler.
+func (s *Scheduler) Start() chan error {
+	// close done when all goroutines created by scheduler have finished
+	done := make(chan struct{})
+	errs := make(chan error)
+
+	// scheduler routine
+	s.wg.Add(1) // decremented by stopAction
+	go func() {
+		for {
+			select {
+			case action := <-s.actions:
+				if err := action.Do(s); err != nil {
+					errs <- err
+				}
+			case err := <-s.runErrs:
+				errs <- err
+			case <-done:
+				close(errs)
+				return
+			}
+		}
+	}()
+
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
+
+	return errs
+}
+
+// RunTask stops the current running task and starts task.
+func (s *Scheduler) RunTask(task Task) {
+	// this guarantees the following:
+	// - if runTaskAction{} is pulled by the scheduler, any errors
+	//   returned by it are reported by the scheduler.
+	// - if Stop() is called _before_ the scheduler picks up this
+	//   action, then this action is skipped.
+	select {
+	case <-s.stopped:
+	case s.actions <- &runTaskAction{
+		task: task,
+	}:
+	}
+}
+
+type runTaskAction struct {
+	task Task
+}
+
+func (r *runTaskAction) Do(s *Scheduler) error {
+	// we no longer care about errors from the old task
+	taskID := s.curTaskID.Add(1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// cancelCtxOnStop calls cancel if Stop() is called before ctx finishes.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-s.stopped:
+			cancel()
+		}
+	}()
+
+	if taskID == 1 {
+		// start the pause container
+		opts := s.pauseRunOptions(r.task)
+		s.run(-1, opts)
+		if err := s.waitForContainerToStart(ctx, opts.ContainerName); err != nil {
+			return fmt.Errorf("wait for pause container to start: %w", err)
+		}
+	} else {
+		// ensure no pause container changes
+		curOpts := s.pauseRunOptions(s.curTask)
+		newOpts := s.pauseRunOptions(r.task)
+		if !maps.Equal(curOpts.EnvVars, newOpts.EnvVars) ||
```

Comment:
Why does envVar / secret changes imply the need to recreate pause?

Reply:
so, like @CaptainCarpensir mentioned [here](https://github.com/aws/copilot-cli/pull/5369#discussion_r1358512919), we don't actually set either for pause containers right now. I added this logic because we will (probably) need to inject IAM credentials to create ssm connections from the pause container. if the value of an env var or secret changes from task to task, we can't create a new pause container with the new values without stopping the existing container and creating a new one.